### PR TITLE
added examples for compat, context and with_context

### PR DESCRIPTION
**Pull request by technetos**
This PR adds examples to the compat, context and with_context methods in result_ext.rs.  

I wasnt sure how much context should be shown in the examples so I tried to keep it as minimal as possible without loosing meaning.

### CODE REVIEW

**Comment by technetos** [comment]
Can I get some advice about why this failed the tests in CI?

File: failure-1.X/src/result_ext.rs [hunk 1]
```diff
@@ -6,14 +6,111 @@ use {Fail, Compat, Context};
 pub trait ResultExt<T, E> {
     /// Wraps the error in `Compat` to make it compatible with older error
     /// handling APIs that expect `std::error::Error`.
+    /// 
+    /// # Example
```
Comment (steveklabnik) [inline failure-1.X/src/result_ext.rs]: convention is to make this `Examples` with an s

**Comment by steveklabnik** [comment]
I've got no idea why the CI fails here though :/

**Comment by withoutboats** [comment]
I'll look into why this is failing on Monday. The dependency set up in failure is pretty complicated and it looks like somehow tests aren't finding the derive crate.

**Comment by withoutboats** [comment]
Actually looking into this now.

**Comment by withoutboats** [comment]
Figured out the issue.

The problem is that your examples use APIs gated by feature flags. When testing, we test under multiple set of feature flags, including `cargo test --no-default-features`, which turns off all feature flags, including those which control `derive(Fail)` and the `Error` type.

I think the best thing to do here is to not run these tests when the `derive` and `std` features are not turned on, but I don't know how to gate running a doc test on certain feature flags. @steveklabnik & @QuietMisdreavus, any ideas?

**Comment by steveklabnik** [comment]
I am not sure that's possible :/

**Comment by withoutboats** [comment]
Do doc tests get the same feature flags that the crate does? One (really hacky) way we could do this is to put the whole thing inside a cfg'd module. That is, inside the doc test do:

```
/// ```
/// # #[cfg(all(feature = "std", feature = "derive"))] mod tests {
/// ...
/// # }
/// ```
```

**Comment by QuietMisdreavus** [comment]
Immediate, hacky suggestion:

```rust
/// ...with the following example:
///
#[cfg_attr(feature = "something", doc = "```")]
#[cfg_attr(not(feature = "something"), doc = "```ignore")]
/// your_code_here();
/// ```
```

As far as i know, rustdoc doesn't have an easy way to turn off doctests based on cfg settings (like cargo features) or even to expose those settings to the doctests themselves. Since rustdoc compiles doctests as external crates, it wouldn't be much use to pass those features on. I would like to double-check that, but as far as i know that option above is the closest i know of. :/

**Comment by withoutboats** [comment]
Misdreavus's suggestion seems better than mine, we could even write a macro possibly?

```rust
/// ...with the following examples:
///
doc_test!("std", "derive")
/// your code here
/// ```
```

**Comment by QuietMisdreavus** [comment]
Alternate suggestion: It's possible that rustdoc may actually pass cfg's to doctests, so you may be able to add something like `# #![cfg(all(feature = "std", feature = "derive"))]` to the beginning of your doctest to apply the cfg to the whole file. I don't remember how rustc reacts to compiling empty inputs (it may complain about the lack of a main function `>_>`) but it's a little bit cleaner than my last suggestion.

**Comment by technetos** [comment]
Alright its a hack but Ive managed to get this working by moving the main out and making the test a separate function like so:
```
    /// ```
    /// # fn main() {
    /// #    tests::run_test();
    /// # }
    /// #
    /// # #[cfg(not(feature = "std"))] mod tests {
    /// # pub fn run_test() {
    /// #
    /// # }
    /// # }
    ///
    /// # #[cfg(feature = "std")] mod tests {
    ...
    /// # pub fn run_test() {
    ///     implementation of doctest here
    /// # }
    ///
    /// # }
    /// ```
```
Is it possible to make this a macro?

**Comment by withoutboats** [comment]
@technetos I think doing this is the right approach. If you can make a macro, that's great, and you can choose where to put it, but its also not necessary if you don't want. :)

**Comment by technetos** [comment]
@withoutboats Ok, for `compat` wrapping the test in a module and separating out main works; however, for `context` and `with_context` the macros that are pulled in for `failure_derive` and `display_derive` still do not work.  

I think the problem lies in the fact that doctests are run as their own crate but I am still working on how to pull in the macros correctly.  @QuietMisdreavus  Do you have any insight or suggestions about this?

**Comment by QuietMisdreavus** [comment]
Looks like you're importing some macros within your `mod tests`, but rustc complains if you import macros from a crate that you're not declaring in your crate root. Moving those crate declarations around to the top of the doctests got them to work for me:

<details><summary>diff that got it to work</summary>

```diff
diff --git a/failure-1.X/src/result_ext.rs b/failure-1.X/src/result_ext.rs
index 894227e..98fcc8a 100644
--- a/failure-1.X/src/result_ext.rs
+++ b/failure-1.X/src/result_ext.rs
@@ -69,6 +69,10 @@ pub trait ResultExt<T, E> {
     /// # Examples
     ///
     /// ```
+    /// # extern crate failure;
+    /// # #[macro_use] extern crate failure_derive;
+    /// # #[macro_use] extern crate display_derive;
+    /// #
     /// # fn main() {
     /// #    tests::run_test();
     /// # }
@@ -80,11 +84,7 @@ pub trait ResultExt<T, E> {
     /// # }
     ///
     /// # #[cfg(all(feature = "std", feature = "derive"))] mod tests {
-    /// # extern crate failure;
-    /// # #[macro_use] extern crate failure_derive;
-    /// # #[macro_use] extern crate display_derive;
-    /// #
-    /// # use tests::failure::ResultExt;
+    /// # use failure::{self, ResultExt};
     /// #
     /// #[derive(Fail, Debug, Display)]
     /// #[display(fmt = "")]
@@ -112,6 +112,10 @@ pub trait ResultExt<T, E> {
     /// # Examples
     ///
     /// ```
+    /// # #[macro_use] extern crate failure;
+    /// # #[macro_use] extern crate failure_derive;
+    /// # #[macro_use] extern crate display_derive;
+    /// #
     /// # fn main() {
     /// #    tests::run_test();
     /// # }
@@ -123,11 +127,7 @@ pub trait ResultExt<T, E> {
     /// # }
     ///
     /// # #[cfg(all(feature = "std", feature = "derive"))] mod tests {
-    /// # #[macro_use] extern crate failure;
-    /// # #[macro_use] extern crate failure_derive;
-    /// # #[macro_use] extern crate display_derive;
-    /// #
-    /// # use tests::failure::ResultExt;
+    /// # use failure::{self, ResultExt};
     /// #
     /// #[derive(Fail, Debug, Display)]
     /// #[display(fmt = "My custom error message")]
```

</details>

**Comment by technetos** [comment]
@steveklabnik @withoutboats @QuietMisdreavus Thanks for the help!

**Pull request merged**